test/stdio/lines.js

```diff
@@ -0,0 +1,94 @@
+import {scheduler} from 'node:timers/promises';
+import test from 'ava';
+import {execa} from '../../index.js';
+import {setFixtureDir} from '../helpers/fixtures-dir.js';
+import {getStdio} from '../helpers/stdio.js';
+
+setFixtureDir();
+
+const bigLine = '.'.repeat(1e5);
+const manyChunks = Array.from({length: 1e5}).fill('.');
+
+const inputGenerator = async function * (input, chunks) {
+	// eslint-disable-next-line no-unused-vars
+	for await (const chunk of chunks) {
+		for (const inputItem of input) {
+			yield inputItem;
+			// eslint-disable-next-line no-await-in-loop
+			await scheduler.yield();
+		}
+	}
+};
+
+const resultGenerator = async function * (lines, chunks) {
+	for await (const chunk of chunks) {
+		lines.push(chunk);
+		yield chunk;
+	}
+};
+
+const textEncoder = new TextEncoder();
+const textDecoder = new TextDecoder();
+
+const stringsToUint8Arrays = (strings, isUint8Array) => isUint8Array
+	? strings.map(string => textEncoder.encode(string))
+	: strings;
+const uint8ArrayToString = (result, isUint8Array) => isUint8Array ? textDecoder.decode(result) : result;
+
+// eslint-disable-next-line max-params
+const testLines = async (t, index, input, expectedLines, isUint8Array) => {
+	const lines = [];
+	const {stdio} = await execa('noop-fd.js', [`${index}`], {
+		...getStdio(index, [
+			inputGenerator.bind(undefined, stringsToUint8Arrays(input, isUint8Array)),
+			resultGenerator.bind(undefined, lines),
+		]),
+		encoding: isUint8Array ? 'buffer' : 'utf8',
+		stripFinalNewline: false,
+	});
+	t.is(uint8ArrayToString(stdio[index], isUint8Array), input.join(''));
+	t.deepEqual(lines, stringsToUint8Arrays(expectedLines, isUint8Array));
+};
+
+test('Split string stdout - n newlines, 1 chunk', testLines, 1, ['aaa\nbbb\nccc'], ['aaa\n', 'bbb\n', 'ccc'], false);
```

Comment:
With those tests, we pass a sets of chunks as `input`, and compare with how it gets split into `expectedLines`.

Under the hood, a first generator emits the `input` chunks. A second generator records each line that it gets as argument. Those `lines` are then compared with `expectedLines`.